File: Miscellaneous/Python/202. Happy Number.py
class Solution:
   def happyNumber(self, n):
        mem = set()
        while n != 1:
            n = sum([int(i) ** 2 for i in str(n)])
            if n in mem:
                return False
            else:
                mem.add(n)
        else:
            return True

obj = Solution()
n = 19
ans = obj.happyNumber(n)
print(ans)

File: Miscellaneous/Golang/202. Happy Number.go
package main

import "fmt"

func main() {

	n := 19
	result := happyNumber(n)
	fmt.Println(result)
}

func happyNumber(n int) bool {
	tempMap := make(map[int]bool)
	tempMap[n] = true

	for {
		sum, x, r := 0, n, 0
		for x > 0 {
			r = x % 10
			sum += r * r
			x /= 10
		}
		exists := tempMap[sum]
		if exists {
			break
		}
		tempMap[n] = true
		n = sum
	}

	if n == 1 {
		return true
	}
	return false
}
